backends/bmv2/pna_nic/pnaNic.cpp

```
@@ -0,0 +1,1046 @@
+        // a prefix including the union name, e.g., "u.h"
+        Util::JsonArray *fields = new Util::JsonArray();
+        for (auto uf : kv.second->to<IR::Type_HeaderUnion>()->fields) {
+            auto uft = typeMap->getType(uf, true);
+            auto h_name = header_name + "." + uf->controlPlaneName();
+            auto h_type = uft->to<IR::Type_StructLike>()->controlPlaneName();
+            unsigned id = ctxt->json->add_header(h_type, h_name);
+            fields->append(id);
+        }
+        ctxt->json->add_union(header_type, fields, header_name);
+    }
+}
+
+void PnaCodeGenerator::createParsers(ConversionContext *ctxt) {
+    {
+        auto cvt = new ParserConverter(ctxt, "main_parser"_cs);
+        auto main_control = parsers.at("main_parser"_cs);
+        main_control->apply(*cvt);
+    }
+}
+
+void PnaCodeGenerator::createExterns() {
+    /* TODO */
+    // add parse_vsets to json
+    // add meter_arrays to json
+    // add counter_arrays to json
+    // add register_arrays to json
+    // add checksums to json
+    // add learn_list to json
+    // add calculations to json
+    // add extern_instances to json
+}
+
+void PnaCodeGenerator::createActions(ConversionContext *ctxt) {
+    auto cvt = new ActionConverter(ctxt, true);
+    for (auto it : actions) {
+        auto action = it.first;
+        action->apply(*cvt);
+    }
+}
+
+void PnaCodeGenerator::createControls(ConversionContext *ctxt) {
+    auto cvt = new BMV2::ControlConverter<Standard::Arch::PNA>(ctxt, "main_control"_cs, true);
+    auto main_control = pipelines.at("main_control"_cs);
+    main_control->apply(*cvt);
+}
+
+void PnaCodeGenerator::createDeparsers(ConversionContext *ctxt) {
+    {
+        auto cvt = new DeparserConverter(ctxt, "main_deparser"_cs);
+        auto main_control = deparsers.at("main_deparser"_cs);
+        main_control->apply(*cvt);
+    }
+}
+
+void PnaCodeGenerator::createGlobals() {
+    /* TODO */
```

I recommend explicit failures. E.g., when there is a construct that is not implemented you should throw an error or `P4C_UNIMPLEMENTED`. This way you do not accidentally compile something that does not work. I recommend that for all the TODOs here.

---

Ok.